intake_xarray/base.py

```diff
@@ -1,38 +1,231 @@
+        and create new coodinates in the output corresponding to pattern
+        fields. If str, is treated as pattern to match on. Default is True.
+    storage_options : dict, optional
+        Any parameters that need to be passed to the remote data backend,
+        such as credentials. For public data on s3 set ``{'anon': True}``.
+    metadata : dict, optional
+        Any additional metadata to include in source.
+    **kwargs :
+        Additional arguments to pass into reader. Only use this or xarray_kwargs.
+    """
+    __doc__ += __inheritted_parameters_doc__
+
+    def __init__(self, urlpath, chunks=None,
+                 concat_dim='concat_dim', merge_dim=None,
+                 xarray_kwargs=None, metadata=None,
+                 path_as_pattern=True, storage_options=None,
+                 reader=None, multireader=None, **kwargs):
+        self.path_as_pattern = path_as_pattern
+        self.urlpath = urlpath
+        self.chunks = chunks
+        self.concat_dim = concat_dim
+        self.merge_dim = merge_dim
+        self.kwargs = xarray_kwargs or kwargs
+        self._ds = None
+        self.storage_options = storage_options
+        self._reader = reader or xr.open_dataset
+        self._multireader = multireader or xr.open_mfdataset
+        super(XarraySource, self).__init__(metadata=metadata)
+
+    def reader(self, filename, chunks, **kwargs):
+        if getattr(self, '_reader', None) is not None:
+            return self._reader(filename, chunks, **kwargs)
+        raise NotImplementedError('Plugin must implement a reader function '
+                                  'which takes at least filename, and chunks '
+                                  'and returns an xarray object')
+
+    def multireader(self, filename, chunks, **kwargs):
+        if getattr(self, '_multireader', None) is not None:
+            return self._multireader(filename, chunks, **kwargs)
+        raise NotImplementedError('Plugin must implement a multireader '
+                                  'function or set this to None. The function '
+                                  'takes a glob representing filenames, '
+                                  'chunks, and arbitrary kwargs and returns '
+                                  'an xarray object')
+
+    def _get_coords(self, das, field_values, index=None):
+        """ Given a list of dask xarray objects and a dict of arrays of coords
+            return a dict of coords that should be set on the concatenated data
+        """
+        import numpy as np
+        return {
+            k: xr.concat(
+                [xr.DataArray(
+                    np.full(das[i].sizes.get(self.concat_dim, 1), v),
+                    dims=self.concat_dim
+                ) for i, v in enumerate(values) if index is None or i in index],
+                dim=self.concat_dim)
+            for k, values in field_values.items() if k != self.merge_dim
+        }
+
+    def _open_files(self, files):
```

Comment:
A lot of logic is in this method. Not sure how to split it up better though. Essentially there are 4 paths through, no pattern and merge_dim, no pattern and concat_dim, pattern and concat_dim, or pattern and merge_dim. The first two are pretty straightforward and then they increase in complexity.

Reply:
Doesn't look too hairy. Perhaps could do with a comment on each branch, saying what it does, and a docstring with your comment above